src/hotspot/share/opto/graphInvariants.cpp

```diff
@@ -0,0 +1,617 @@
+      for (DUIterator_Fast imax, i = center->fast_outs(imax); i < imax; i++) {
+        Node* out = center->fast_out(i);
+        ss.print("  ");
+        out->dump("\n", false, &ss);
+      }
+      return false;
+    }
+    return true;
+  }
+  const uint _expect_outcnt;
+};
+
+struct AtSingleOutputOfType : Pattern {
+  AtSingleOutputOfType(bool (Node::*type_check)() const, const Pattern* pattern) : _type_check(type_check), _pattern(pattern) {
+  }
+
+  bool check(const Node* center, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    Node_List outputs_of_correct_type;
+    for (DUIterator_Fast imax, i = center->fast_outs(imax); i < imax; i++) {
+      Node* out = center->fast_out(i);
+      if ((out->*_type_check)()) {
+        outputs_of_correct_type.push(out);
+      }
+    }
+    if (outputs_of_correct_type.size() != 1) {
+      ss.print_cr("Non-unique output of expected type. Found: %d.", outputs_of_correct_type.size());
+      for (uint i = 0; i < outputs_of_correct_type.size(); ++i) {
+        outputs_of_correct_type.at(i)->dump("\n", false, &ss);
+      }
+      return false;
+    }
+    bool result = _pattern->check(outputs_of_correct_type.at(0), steps, path, ss);
+    if (!result) {
+      steps.push(outputs_of_correct_type.at(0));
+      path.push(LocalGraphInvariant::OutputStep);
+    }
+    return result;
+  }
+  bool (Node::*_type_check)() const;
+  const Pattern* const _pattern;
+};
+
+struct PatternBasedCheck : LocalGraphInvariant {
+  const Pattern* const _pattern;
+  explicit PatternBasedCheck(const Pattern* pattern) : _pattern(pattern) {}
+  CheckResult check(const Node* center, LazyReachableCFGNodes& reachable_cfg_nodes, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    return _pattern->check(center, steps, path, ss) ? CheckResult::VALID : CheckResult::FAILED;
+  }
+};
+
+struct IfProjections : PatternBasedCheck {
+  IfProjections()
+      : PatternBasedCheck(
+            And::make(
+                new HasNOutputs(2),
+                new AtSingleOutputOfType(&Node::is_IfTrue, new True()),
+                new AtSingleOutputOfType(&Node::is_IfFalse, new True()))) {
+  }
+  const char* name() const override {
+    return "IfProjections";
+  }
+  CheckResult check(const Node* center, LazyReachableCFGNodes& reachable_cfg_nodes, Node_List& steps, GrowableArray<int>& path, stringStream& ss) const override {
+    if (!center->is_If()) {
+      return CheckResult::NOT_APPLICABLE;
+    }
+    CheckResult r = PatternBasedCheck::check(center, reachable_cfg_nodes, steps, path, ss);
+    if (r == CheckResult::FAILED) {
+      if (reachable_cfg_nodes.is_node_dead(center)) {
+        // That's ok for dead nodes right now. It might be too expensive to collect for IGVN, but it will be removed in loop opts.
+        ss.reset();
+        return CheckResult::VALID;
+      }
+    }
+    return r;
```

Comment:
Also this could probably be handled with a pattern wrapping mechanism, right?
`FailOnlyForLiveNodes( <the pattern from above> )`

Reply:
That would be possible. It's still rare, and I'm not convinced we should make so specialized such patterns for one usecase. If it gets more usage, sure, that would be something to do. The only other usage is not so easy to phrase as template.